### PR TITLE
Allow to box messages in the channels implicitly

### PR DESCRIPTION
**Pull request by vstakhov**
This PR allows to box all messages when sending them over channels. The behaviour is controlled by attribute `boxed_messages` of the `orchestra` macro, for example:

~~~rust
#[orchestra(signal=SigSigSig, event=Event, gen=AllMessages, error=OrchestraError, boxed_messages=true)]
pub struct Orchestra {
	#[subsystem(consumes: MsgA, sends: [MsgB])]
	sub_a: AwesomeSubSysA,

	#[subsystem(consumes: MsgB, sends: [MsgA])]
	sub_b: AwesomeSubSysB,
}
~~~

This change is intended to be transparent for the orchestrated subsystems.

### CODE REVIEW

**Comment by drahnr** [comment]
I think the approach would work, my question would be it might be less of a :hammer: to have it on a per message type, such that one could either annotate the message enum that a subsystem consumes with a proc-macro / individual fields / the subsystem;

```
#[orchestra::giftwrap(all)]
enum FooMessage {
}

#[orchestra::giftwrap]
enum BarMessage {

#[ribbon]
X(Fat),
}
```

(obviously the naming has to be adjusted)

or
 
```
#[orchestra(signal=SigSigSig, event=Event, gen=AllMessages, error=OrchestraError, boxed_messages=true)]
pub struct Orchestra {
	#[subsystem(consumes: boxed MsgA, sends: [MsgB])]
	sub_a: AwesomeSubSysA,

	#[subsystem(consumes: boxed MsgB, sends: [MsgA])]
	sub_b: AwesomeSubSysB,
}
```

such that the consumer defines the behavior of transparent boxing

**Comment by vstakhov** [comment]
Benchmarks from my laptop (osx):

no jemalloc:

```
send boxed message u8   time:   [320.24 ns 328.55 ns 336.36 ns]
send unboxed message u8 time:   [285.09 ns 291.20 ns 296.45 ns]
send boxed message 1029 bytes time:   [738.99 ns 763.95 ns 788.70 ns] 
send unboxed message 1029 bytes time:   [615.58 ns 621.74 ns 628.62 ns]
```

preloaded jemalloc:

```
send boxed message u8   time:   [289.96 ns 297.19 ns 303.40 ns]
send unboxed message u8 time:   [252.18 ns 257.18 ns 261.48 ns]
send boxed message 1029 bytes time:   [386.75 ns 401.52 ns 415.99 ns]
send unboxed message 1029 bytes time:   [613.14 ns 619.94 ns 627.32 ns]
```

So almost no difference without jemalloc (boxed version is slightly slower in all cases) and significant difference when jemalloc is loaded.

File: orchestra/Cargo.toml [hunk 1]
```diff
@@ -11,7 +11,7 @@ autoexamples = false
 
 [dependencies]
 tracing = "0.1.35"
-futures = "0.3"
+futures = { version = "0.3", features = ["thread-pool"] }
```
Comment (drahnr) [inline orchestra/Cargo.toml:14]: Where do we require this?

**Review by drahnr — CHANGES_REQUESTED** [review]
One dependency question.

Do we want to have an API to send i.e. a pre-boxed version of some messages?

Other than that LGTM (although I am not the greatest fan of the added complexity to orchestra. I'd have preferred to make it around the message types being sent, so orchestra can stay oblivious to the message type )

**Review by drahnr — APPROVED** [review]

**Pull request merged**